### PR TITLE
Sketch mini-pyro implementation

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #1451, #1554
Pair coded with @martinjankowiak 

This implements "Pyro in 150 lines" including an example using SVI to fit a normal-normal model.

My objective is to create a minimal example, and then add some deeper poutine usage in subsequent PRs. I'd like to keep total line count under 500, and add documentation as needed. In lieu of a hackathon, I suggest all Pyro team members fit minified versions of their ideas into this example (in separate PRs).

## Questions for reviewers

- [x] @eb8680 can you suggest minimal changes to this poutine stuff? I'd be happy merging this first version and letting you take a second pass.

## Tested

- [x] added to test/test_examples.py

### CODE REVIEW

File: examples/mini_pyro.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    def __init__(self, model, guide, optim, loss):
+        self.model = model
+        self.guide = guide
+        self.optim = optim
+        self.loss = loss
+
+    def step(self, *args, **kwargs):
+        loss = self.loss(self.model, self.guide, *args, **kwargs)
+        loss.backward()
+        params = list(PARAM_STORE.values())
```
Comment (eb8680) [inline examples/mini_pyro.py]: We should only grab params that appear in `loss`

File: examples/mini_pyro.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            return self.fn(*args, **kwargs)
+
+
+class trace(Messenger):
+    def __enter__(self):
+        self.trace = OrderedDict()
+        return super(trace, self).__enter__()
+
+    def process_msg(self, msg):
+        self.trace[msg["name"]] = msg
```
Comment (eb8680) [inline examples/mini_pyro.py]: `self.trace[msg["name"]] = msg.copy()`

**Comment by eb8680** [comment]
Also, in terms of organization, maybe we should make this into a notebook and automatically extract a script with `nbconvert`?

**Comment by fritzo** [comment]
> maybe we should make this into a notebook

While I like the idea of publishing this as a tutorial that shows up on docs.pyro.ai, I find it frustrating to edit, refactor, test, profile, and import jupyter notebooks. @jpchen Is there a way we can make a nice html page out of this file?

**Comment by jpchen** [comment]
probably. i think the tutorials webpage is a good home for it. not sure how well it plays with sphinx as a standalone html. at the very least it can be exported into a one-cell notebook using an `nbformat` script

**Comment by fritzo** [comment]
> missing ... block, plate/markov, and parallel enumeration.

Yeah I'd really like to see `block`, since users seem most confused about poutines and messengers. I feel like `plate` might be useful to see how broadcasting works, but I think this file will be most helpful if we focus on the nonstandard-interpretation aspects and leave tensor shape explanation for elsewhere. Similarly, enumeration would be nice to demonstrate nonstandard interpretation, but I think it's worth implementing a limited version that assumes a single plate nesting, so as to focus on the PL aspects.

**Comment by eb8680** [comment]
> users seem most confused about poutines and messengers

I guess it depends on the goals of this example - if the goal is to explain effect handlers, then that's fine, but if the goal is to explain how Pyro works then it seems like many more people will want to know how `TraceEnum_ELBO` works, since that's currently our main selling point and most useful feature.  That means at least explaining `plate` and parallel enumeration via broadcasting and `ubersum`.

**Comment by martinjankowiak** [comment]
my two cents is that this should mostly focus on examples of non-standard interpretation stuff and that 2-3 instantiations of that concept should be sufficient. it'd be a shame if too much complexity scared people off or confused them. if we want tutorials that explain how, e.g., traceenum works, i think that should go elsewhere.

**Comment by eb8680** [comment]
> my two cents is that this should mostly focus on examples of non-standard interpretation stuff and that 2-3 instantiations of that concept should be sufficient. it'd be a shame if too much complexity scared people off or confused them

I think you're probably right that building all the way up to enumeration would be too much material, but on the other hand, without a meatier example of some kind as payoff at the end I'm not sure who this tutorial would be useful/interesting for.  What do you think we should build up to instead?
